src/lib.rs

```diff
@@ -169,6 +178,36 @@ macro_rules! impl_single {
                 *self = serde_json_core::from_slice(value)?.0;
                 Ok(())
             }
+
+            // This implementation is the base case for primitives where it will
+            // yield once for self, then return None on subsequent calls.
+            fn recursive_iter<const TS: usize, const VS: usize>(
+                &self,
+                index: &mut [usize],
+                _topic: &mut heapless::String<TS>,
+                value: &mut heapless::String<VS>,
+            ) -> Option<()>
```

Comment:
Instead of pass-by-reference, just have the prototype for `recursive_iter()` construct the `topic` and `value` internally. No reason to pass it as input from what I can see.

Also, `value` should be a `Vec<u8>`. `String` is a UTF-8 encodeded string, and we make no guarantees about the serialization format of the underlying data right now.

Reply:
Ah wait, I now see why they're passed as input arguments. That's because we're constantly appending to `topic`. That being said, `value` could be returned as the result instead - I don't see why that needs to be an input argument

Reply:
Or alternatively take `value` as a mutable slice reference so users can manage memory for it and we lose a template argument